src/client/clientmap.cpp

```diff
@@ -610,3 +659,243 @@ void ClientMap::PrintInfo(std::ostream &out)
+
+				u32 c = mesh->getMeshBufferCount();
+				for (u32 i = 0; i < c; i++) {
+					scene::IMeshBuffer *buf = mesh->getMeshBuffer(i);
+
+					video::SMaterial &material = buf->getMaterial();
+					video::IMaterialRenderer *rnd =
+							driver->getMaterialRenderer(
+									material.MaterialType);
+					bool transparent = (rnd && rnd->isTransparent());
+					if (transparent == is_transparent_pass) {
+						drawbufs.add(mesh->getMeshBuffer(i),
+								block_pos, layer);
+					}
+				}
+			}
+		}
+	}
+
+	TimeTaker draw("Drawing Shadows mesh buffers");
+
+	core::matrix4 m; // Model matrix
+	v3f offset = intToFloat(m_camera_offset, BS);
+
+	// Render all layers in order
+	for (auto &lists : drawbufs.lists) {
+		for (MeshBufList &list : lists) {
+			// Check and abort if the machine is swapping a lot
+			if (draw.getTimerTime() > 1000) {
+				infostream << "ClientMap::renderMapShadows(): Rendering "
+					      "took "
+					      ">2s, "
+					   << "returning." << std::endl;
+				return;
+			}
+
+			for (auto &pair : list.bufs) {
+				scene::IMeshBuffer *buf = pair.second;
+				video::SMaterial local_material = buf->getMaterial();
+				video::IMaterialRenderer *rnd = driver->getMaterialRenderer(local_material.MaterialType);
+				bool transparent = (rnd && rnd->isTransparent());
+				if (transparent == is_transparent_pass) {
+					local_material.MaterialType = material.MaterialType;
+					local_material.BackfaceCulling = material.BackfaceCulling;
+					local_material.FrontfaceCulling =	material.FrontfaceCulling;
+					local_material.Lighting = false;
+					
+				}
+				driver->setMaterial(local_material);
+				v3f block_wpos = intToFloat(
+						pair.first * MAP_BLOCKSIZE, BS);
+				m.setTranslation(block_wpos - offset);
+
+				driver->setTransform(video::ETS_WORLD, m);
+				driver->drawMeshBuffer(buf);
+				vertex_count += buf->getVertexCount();
+			}
+			
+			drawcall_count += list.bufs.size();
+		}
+	}
+
+	g_profiler->avg(prefix + "draw Shadow meshes [ms]", draw.stop(true));
+	g_profiler->avg(prefix + "vertices drawn [#]", vertex_count);
+	g_profiler->avg(prefix + "drawcalls [#]", drawcall_count);
+}
+
+
+/*
+Custom update draw list for the pov of shadow light.
+@Liso double or triple check this fn and optimize it.
+*/
+void ClientMap::updateDrawListShadow(
+		v3f shadow_light_pos, v3f shadow_light_dir, float shadow_range)
+{
+	ScopeProfiler sp(g_profiler, "CM::updateDrawListShadow()", SPT_AVG);
+
+	v3f camera_position = shadow_light_pos;
+	v3f camera_direction = shadow_light_dir;
+	// I "fake" fov just to avoid creating a new function to handle orthografic
+	// projection.
+	f32 camera_fov = m_camera_fov * 1.5f;
+
+	v3s16 cam_pos_nodes = floatToInt(camera_position, BS);
+	v3s16 p_blocks_min;
+	v3s16 p_blocks_max;
+	getBlocksInViewRangeForShadow(cam_pos_nodes, &p_blocks_min, &p_blocks_max, shadow_range);
+
+
+	std::vector<v2s16> blocks_in_range;
+	 
+	clean_mblock_shadows();
+
+	if (m_drawlist_shadow.empty())
+		m_drawlist_shadow=m_drawlist;
```

Comment:
This will always copy the normal drawlist to the shadow drawlist.
1. Why? The rest of the method is supposed to fill it.
2. Since `refDrop()` will be called when the (respective) list is cleared by either `updateDrawList` or `updateDrawListShadow` duplicating blocks here will disturb the reference counting.

Reply:
Yes, this is because the first version I did it got the clientmap blocks and It searched for the block before adding it, but it turns to be slower that adding them without checking.
This code, must be removed.

Tnx :)